libmscore/instrument.cpp

```diff
@@ -655,12 +667,10 @@ void Channel::write(XmlWriter& xml, const Part* part) const
             if (e.type() == ME_INVALID)
                   continue;
             if (e.type() == ME_CONTROLLER) {
-                  // don't write if automatically switched
-                  if ((e.dataA() == CTRL_HBANK || e.dataA() == CTRL_LBANK) &&  !_userBankController)
+                  // Don't write bank if automatically switched, but always write if switched by the user
+                  if (e.dataA() == CTRL_HBANK && e.dataB() == 0 && !_userBankController)
                         continue;
-                  if (e.dataA() == CTRL_HBANK && e.dataB() == 0)
-                        continue;
-                  if (e.dataA() == CTRL_LBANK && e.dataB() == 0)
+                  if (e.dataA() == CTRL_LBANK && e.dataB() == 0 && !_userBankController)
```

Comment:
It looks like this part of the patch (the changes inside `if (e.type() == ME_CONTROLLER)`) is not really necessary to fix the issue and causes those extra `<controller>` tags to appear. At least in my tests the scenario from the [issue 290121](https://musescore.org/en/node/290121) works correctly without changing this part of the code. Are there any reasons why this change might be necessary to do?

Reply:
This is just a reworking of the initial statement, I guess, with the added check for if it is 0 or not. I think it should be alright without these changes then... but I'd need to test it.

Reply:
No hang on, we _always_ want to write the bank controller if it was set by the user, even if the data is 0, so that if it's set to 0 from a number of 1, say, it's not automatically switched back to 1 when its read (since if a bank value is read from the file, then we don't automatically switch to expressive patches). So, this line is correct and should stay as it is.

The old line doesn't account for this since if the bank is 0, then it won't be written, even if the user has changed it to that.

Reply:
Then it does not agree with [the other part](https://github.com/musescore/MuseScore/pull/5094/files#diff-9c93d8fdd487e712679e845aba7e30b6R2500) of this patch. If you explicitly write only one of HBANK or LBANK controllers, the value of the other will be 0 by default, and `_userBankController` flag will be set to true, so the effect will be essentially the same. If you explicitly write both of them and their value is 0, then `_userBankController` flag will be anyway reset to false. So, if I am correct, there are no situations where writing zero values for those controllers makes any difference to the result being obtained on file reading.

Reply:
Hmmm... it's worth noting that that other part of the patch is just a slightly hacky fix to get SND working with 114 import. But yes, I think you might be right. I'll change it back.

Reply:
Ok, testing it, it actually _is_ important to have it the way it is. With your proposed changes:

1. create a score with an expressive instrument
2. change it to the non expr version in the mixer
3. save and close
4. reopen the file
5. it's changed back to the expr version.

This happens for the reasons above:



> we _always_ want to write the bank controller if it was set by the user, even if the data is 0, so that if it's set to 0 from a number of 1, say, it's not automatically switched back to 1 when its read (since if a bank value is read from the file, then we don't automatically switch to expressive patches

Hope that makes sense, I realise I'm not explaining it the best.

Reply:
Ah, I didn't notice that chunk of code actually belongs to `read114`. Sorry then, it turns out there is no contradiction here indeed, we indeed are better to write those tags.

Reply:
It's fine, it's a really confusing area of the codebase unfortunately. Thanks for merging :)